frontend/test/lit/test_multi_qubit_gates.py

```diff
@@ -55,7 +55,7 @@ def circuit():
         ]
     )
     # CHECK: {{%.+}} = "quantum.unitary"({{%.+}}, {{%.+}}, {{%.+}}) : (tensor<4x4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
-    qml.QubitUnitary(U2, wires=[1, 2])
+    qml.QubitUnitary(U2, wires=[0, 2])
```

Comment:
No change required, just wondering if there was a bug in the test before?

Reply:
I think we could call it an error in test, but it depends on our assumptions regarding the hardware. In the original version of the test the first quantum op affected wire 0 and the second one affected wires 1 and 2 (and btw we are measuring wire 0 only). Given the jax sorting, there is no good way to force the final program to match the littest's expectations. In this particular case the algorithm placed the second op before the first one. So I changed the wire to make the order to really match the lit-rules we have here.

Reply:
I think we could call it an error in test, but it depends on our assumptions regarding the hardware. In the original version of the test the first quantum op affected wire 0 and the second one affected wires 1 and 2 (and btw we are measuring wire 0 only). Given the jax sorting, there is no good way to force the final program to match the littest's expectations. In this particular case the algorithm placed the second op before the first one. So I changed the wire to make the order to really match the lit-rules we have here.

Reply:
So you are saying essentially that `qml.QubitUnitary` is dead code and the compiler may rearrange it to be placed even after the deallocation? Or only after the measurement?

Reply:
The deallocation by itself is not a problem because I made precautions to always include it as a last operation. But if we have
```
b,d = ..
a = f1(b)
c = f2(d)
use(a,c)
```
then the compiler will arrange f1 and f2 according to some deterministic but unspecified logic. Currently this does not always match the test expectations.

Reply:
I am thinking how to specify this logic. It could be not _always_ possible due to mid-circuit measurements, but maybe we can improve it for other cases

Reply:
I think if it is dead code, it doesn't matter if it is re-arranged. We should probably optiimze it and remove it from the computation. (Potentially issue a warning if dead code is found as the user might not have intended it). But this is something beyond the scope of this PR.

Reply:
Just looking at the issue of testing the code, LIT allows you to specify checks without a particular order: `CHECK-DAG`
It can't always be used depending on the complexity of the of test, but here it should work.

I'm glad the test uncovered the question of rearranging code however. To be honest I'm not a huge fan that it is happening at such an early stage. From a functional point of view it is correct, but thinking about debugging code by looking at the jaxpr or initial mlir it is much easier to map from IR to source if the order were preserved. Once we run optimizations I think rearrangements are fair game.
Surprisingly the original tracing implementation managed to preserve the order rather well.

Reply:
Yes, speaking about the sorting - I believe it is correct to think that we split the original flow of instructions into two lists: classical instructions go directly to `frame.eqn` and quantum instructions go to the nested quantum tape. After that we merge the lists and at this point (in the current implementation) we indeed miss some information, e.g. line numbers to merge them precisely. I am pretty sure I can collect and preserve all the information to keep the original line order. Can do it if we want.

> Surprisingly the original tracing implementation managed to preserve the order rather well

The core problem we solve now is the list merging. The current implementation "adds" quantum tape (after the transformations!) to the already existing list of classical equation and then sort everything based on the properties of items - the variables they produce/consume. Can we think of some other way? I will think..

Reply:
Yes I think you are right. I guess the original implementation preserved the order by maintaining the sequence of values that connect the classical & quantum parts.

> line numbers - which is required for the precise merging
Maybe we could open a GitHub issue about it as a future improvement? That way we can still go ahead with the PR.

Reply:
Created this one https://github.com/PennyLaneAI/catalyst/issues/284

Reply:
Thanks :)